From e3560d3a3fb3818011987502d286e2a48d8bca36 Mon Sep 17 00:00:00 2001
From: Lawrence Rust <lvr@softsystem.co.uk>
Date: Thu, 3 Mar 2016 17:17:13 +0000
Subject: [PATCH 307/348] MythMusic: Prevent corruption of ripped CD music
 files

When an ogg/flac file is created from a ripped track, meta
information is written in the Vorbis/Flac/Encoder dtor.
However, the meta information is written before the output file data
is flushed.  This can cause corruption of the last few frames.

Signed-off-by: Lawrence Rust <lvr@softsystem.co.uk>
---
 mythplugins/mythmusic/mythmusic/flacencoder.cpp   |   11 +++++++++++
 mythplugins/mythmusic/mythmusic/flacencoder.h     |    1 +
 mythplugins/mythmusic/mythmusic/vorbisencoder.cpp |    6 ++++++
 3 files changed, 18 insertions(+)

diff --git a/mythplugins/mythmusic/mythmusic/flacencoder.cpp b/mythplugins/mythmusic/mythmusic/flacencoder.cpp
index f635006..7fd5895 100644
--- a/mythplugins/mythmusic/mythmusic/flacencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/flacencoder.cpp
@@ -29,6 +29,12 @@ FlacEncoder::FlacEncoder(const QString &outfile, int qualitylevel,
                          MusicMetadata *metadata)
            : Encoder(outfile, qualitylevel, metadata)
 {
+    if (m_out)
+    {
+        fclose(m_out);
+        m_out = NULL;
+    }
+
     sampleindex = 0;
 
     bool streamable_subset = true;
@@ -75,6 +81,8 @@ FlacEncoder::FlacEncoder(const QString &outfile, int qualitylevel,
         LOG(VB_GENERAL, LOG_ERR,
             QString("Error initializing FLAC encoder. Got return code: %1")
                 .arg(ret));
+        encoder_delete(encoder);
+        encoder = NULL;
     }
 
     for (int i = 0; i < NUM_CHANNELS; i++)
@@ -101,6 +109,9 @@ FlacEncoder::~FlacEncoder()
 
 int FlacEncoder::addSamples(int16_t *bytes, unsigned int length)
 {
+    if (!encoder)
+        return EENCODEERROR;
+
     unsigned int index = 0;
 
     length /= sizeof(int16_t);
diff --git a/mythplugins/mythmusic/mythmusic/flacencoder.h b/mythplugins/mythmusic/mythmusic/flacencoder.h
index 3112a4d..7002482 100644
--- a/mythplugins/mythmusic/mythmusic/flacencoder.h
+++ b/mythplugins/mythmusic/mythmusic/flacencoder.h
@@ -85,6 +85,7 @@ class FlacEncoder : public Encoder
     FlacEncoder(const QString &outfile, int qualitylevel, MusicMetadata *metadata);
    ~FlacEncoder();
     int addSamples(int16_t *bytes, unsigned int len);
+    virtual bool isValid() { return !!encoder; }
 
   private:
     FLAC_ENCODER *encoder;
diff --git a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
index 3955e86..ad55599 100644
--- a/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
+++ b/mythplugins/mythmusic/mythmusic/vorbisencoder.cpp
@@ -98,6 +98,12 @@ VorbisEncoder::VorbisEncoder(const QString &outfile, int qualitylevel,
 VorbisEncoder::~VorbisEncoder()
 {
     addSamples(nullptr, 0); //flush
+    if (m_out)
+    {
+        fclose(m_out);
+        m_out = NULL;
+    }
+
     ogg_stream_clear(&os);
     vorbis_dsp_clear(&vd);
     vorbis_block_clear(&vb);
-- 
1.7.9.5

